### PR TITLE
Network fixes: reduce communication, show traffic, ban for abusive behavior

### PR DESCRIPTION
**Pull request by ilblackdragon**
For each peer track how many bytes in/out, and last blocks received/requested.
Use that info to reduce communication and show total stats in logs on traffic.
Additionally keep track of the latest chain of peers based on communication received to finish caching when catch up is done.
Also if peer sends too many messages within a period of time - ban it for abuse.

### CODE REVIEW

File: chain/network/src/peer.rs [hunk 1]
```diff
@@ -61,10 +111,19 @@ impl Peer {
...
         }
     }
 
     fn send_message(&mut self, msg: PeerMessage) {
+        // Skip sending block and headers if we received it or header from this peer.
+        // Record block requests in tracker.
+        match msg {
+            PeerMessage::Block(b) if self.tracking.has_received(b.hash()) => return,
+            PeerMessage::BlockHeaderAnnounrce(h) if self.tracking.has_received(h.hash()) => return,
+            PeerMessage::BlockRequest(h) => self.tracker.push_request(h),
```
Comment (bowenwang1996) [inline chain/network/src/peer.rs]: When it sends blocks, why doesn't it push the hash to tracker?

File: chain/network/src/peer.rs [hunk 2]
```diff
@@ -107,9 +166,13 @@ impl Peer {
         let network_client_msg = match msg {
             PeerMessage::Block(block) => {
                 // TODO: add tracking of requests here.
```
Comment (bowenwang1996) [inline chain/network/src/peer.rs]: Remove this todo?

File: chain/network/src/peer.rs [hunk 3]
```diff
@@ -107,9 +166,13 @@ impl Peer {
...
-                NetworkClientMessages::Block(block, peer_id, false)
+                let block_hash = b.hash();
+                self.tracker.push_recevied(block_hash);
+                NetworkClientMessages::Block(block, peer_id, self.tracker.has_request(block_hash))
             }
             PeerMessage::BlockHeaderAnnounce(header) => {
+                let block_hash = b.hash();
+                self.tracker.push_recevied(block_hash);
```
Comment (bowenwang1996) [inline chain/network/src/peer.rs]: Why only pushing header and block to tracker?

File: chain/network/src/peer.rs (hunk 1 above)
Comment (ilblackdragon) [reply chain/network/src/peer.rs]: Because it's not a request or recevied?

File: chain/network/src/peer.rs (hunk 3 above)
Comment (ilblackdragon) [reply chain/network/src/peer.rs]: Transaction tracking will be separate as part of the "tx routing" #1038

**Pull request merged**